src/app/services/xml-parsers/structure-xml-parser.service.ts

```diff
@@ -21,63 +21,121 @@ export class StructureXmlParserService {
-        const page: Page = {
-          id: `page_${new Date().getTime()}`,
-          label: 'Main Text',
-          originalContent: content as XMLElement[],
-          parsedContent: content.map(child => this.genericParserService.parse(child as XMLElement))
-            // tslint:disable-next-line: no-string-literal
-            .filter(c => !!c['content']), // TODO: FIXME: fix property access
-        };
-        pages[page.id] = page;
-        pagesIndexes.push(page.id);
+        this.parseDocument(document, pages, pagesIndexes);
       }
-      console.log(pages);
+
+      console.log('### PAGES ###', pages);
     }
 
     return {
       pages,
       pagesIndexes,
     };
   }
+
+  parseDocumentPages(pageElements: NodeListOf<Element>, pagesNumber: number, pages: Map<Page>, pagesIndexes: string[]) {
+    pageElements.forEach((element, i) => {
+      let pageContent: XMLElement[] = [];
+      const isPbFront = isNestedInElem(element, 'front');
+
+      if (pagesNumber === 1) {
+        pageContent = getElementsAfterTreeNode(element);
+      }
+
+      if (pagesNumber !== 1 && !isPbFront) {
+        const isFirstPage = i === 0;
+        const isLastPage = i === pagesNumber - 1;
+
+        if (isFirstPage) {
+          pageContent = getElementsBetweenTreeNode(element.closest('body'), pageElements[i + 1]);
+        }
+
+        if (isLastPage) {
+          pageContent = getElementsAfterTreeNode(element);
+        }
+
+        if (!isFirstPage && !isLastPage) {
+          pageContent = getElementsBetweenTreeNode(element, pageElements[i + 1]);
+        }
+      }
+
+      if (isPbFront) {
+        pageContent = getElementsBetweenTreeNode(element, pageElements[i + 1])
+          .filter((el: XMLElement) => el.nodeType !==8 && this.getFrontOriginalElements(el));
+      }
+
+      // Exclude nodes in <back>
+      pageContent = pageContent.filter(el => {
+        if (el.nodeType === 3) {
+          return !el.parentElement.closest('back') && el.parentElement.tagName !== 'back';
+        }
+        if (el.nodeType === 1) {
+          return !el.closest('back') && el.tagName !== 'back';
+        }
+
+        return false;
+      });
+
+      const page: Page = {
+        id: element.getAttribute('xml:id') || 'page_' + (pagesIndexes.length + 1),
+        label: element.getAttribute('n') || 'Page ' + (pagesIndexes.length + 1),
+        originalContent: pageContent,
+        parsedContent: this.parsePageContent(isPbFront, pageContent),
+      };
+      pages[page.id] = page;
+      pagesIndexes.push(page.id);
+    });
+  }
+
+  parseDocument(document: XMLElement, pages: Map<Page>, pagesIndexes: string[]) {
+    // No <pb> used => TODO: Decide how to handle text division
+    console.warn('TODO: Decide how to handle text division when there are no <pb>s');
+    const mainText = document.querySelector('text');
+    const content = Array.from(mainText.childNodes);
+    const page: Page = {
+      id: `page_${new Date().getTime()}`,
```

Comment:
Is there a `getId` function elsewhere?
If yes use that, otherwise make one and use it

Reply:
I wouldn't use the current time to define the id of pages without identification. In this way, every time you load the edition the id changes and you cannot load specific pages based on URL parameters. I would use the xpath, or just the position/count of the element

Reply:
The id will probably go on the URL. Is the current time the best way to define the id of "virtual" pages?